### ISSUE TITLE
SQLTransientConnectionException

### ISSUE THREAD

**Issue by mibby** [Esophose/PlayerParticles#35]
Java 13 (OpenJDK)
Purpur dev 486 (Paper Fork / Spigot 1.15.2)
PlayerParticles dev 14 (https://ci.codemc.io/job/Esophose/job/PlayerParticles/changes)

Since updating to the latest dev build and to java 13, I seem to be getting this error at start-up. Did something change recently breaking mysql tables or the database? The fixed particle I set a while ago at my spawn location is still loading fine.

I try to update dev builds / compile the source (prior to jenkins) every couple days if there are new commits. 

```
[10:01:45] [Craft Scheduler Thread - 27/ERROR]: [PlayerParticles] An error occurred executing a MySQL query: HikariPool-1 - Connection is not available, request timed out after 30000ms.
[10:01:45] [Craft Scheduler Thread - 4/ERROR]: [PlayerParticles] An error occurred executing a MySQL query: HikariPool-1 - Connection is not available, request timed out after 30010ms.
[10:01:45] [Craft Scheduler Thread - 27/WARN]: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:676)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:190)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:155)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at dev.esophose.playerparticles.database.MySQLConnector.connect(MySQLConnector.java:41)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$getPPlayer$2(DataManager.java:112)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftTask.run(CraftTask.java:84)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[10:01:45] [Craft Scheduler Thread - 27/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:830)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30010ms.
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:676)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:190)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:155)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.database.MySQLConnector.connect(MySQLConnector.java:41)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$saveParticleGroup$10(DataManager.java:313)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.async(DataManager.java:533)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.saveParticleGroup(DataManager.java:313)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$getPPlayer$1(DataManager.java:244)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.database.MySQLConnector.connect(MySQLConnector.java:42)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$getPPlayer$2(DataManager.java:112)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftTask.run(CraftTask.java:84)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[10:01:45] [Craft Scheduler Thread - 4/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:830)
[10:01:45] [Craft Scheduler Thread - 30/ERROR]: [PlayerParticles] An error occurred executing a MySQL query: HikariPool-1 - Connection is not available, request timed out after 30000ms.
[10:01:45] [Craft Scheduler Thread - 30/WARN]: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:676)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:190)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:155)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.database.MySQLConnector.connect(MySQLConnector.java:41)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$saveParticleGroup$10(DataManager.java:313)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.async(DataManager.java:533)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.saveParticleGroup(DataManager.java:313)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$getPPlayer$1(DataManager.java:244)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.database.MySQLConnector.connect(MySQLConnector.java:42)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at dev.esophose.playerparticles.manager.DataManager.lambda$getPPlayer$2(DataManager.java:112)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftTask.run(CraftTask.java:84)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at org.bukkit.craftbukkit.v1_15_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[10:01:45] [Craft Scheduler Thread - 30/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:830)
```

Other plugins that rely on the same mysql database connect and load fine without timing out or erroring. :( The SQL database is local as well, no remote access.

**Comment by Esophose** [comment]
I've had one other person message me with this error recently. I haven't made any changes to the MySQL in a long while. I'm still entirely unsure what might be causing it and will have to do some testing. Let me know if you're able to find the cause of it.

**Comment by Esophose** [comment]
Could you try the following build?: [https://ci.codemc.io/job/Esophose/job/PlayerParticles/15/](https://ci.codemc.io/job/Esophose/job/PlayerParticles/15/)

Turns out I actually did make a database change during the v7.0 update to avoid creating async threads where unnecessary while saving data. Turns out this was a _bad idea_ since it was causing multiple connections to be potentially opened at the same time, and if max connections were reached, it would just deadlock. I am unsure how I didn't run into this during testing and don't know why it wasn't happening with SQLite. 

Please let me know if this happens to fix the issue.

**Comment by mibby** [comment]
So far it seems good and fixed, thanks! :)

**Comment by Esophose** [comment]
Thanks for the confirmation :+1:
Going to go ahead and close this issue. The change will be released in v7.5 whenever I feel the update is ready.
